src/mixins/java/org/spongepowered/common/mixin/core/server/level/PlayerSpawnFinderMixin.java

```diff
@@ -0,0 +1,160 @@
+        throw new IllegalStateException();
+    }
+
+    /**
+     * A mix of loop and recursion.
+     * Loop is preferred when possible.
+     *
+     * @author Yeregorix
+     * @reason Fix <a href="https://bugs.mojang.com/browse/MC/issues/MC-304426">MC-304426</a>.
+     */
+    @Overwrite
+    private void scheduleNext() {
+        int i;
+        while ((i = this.nextCandidateIndex++) < this.candidateCount) {
+            if (this.finishedFuture.isDone()) {
+                return;
+            }
+
+            final int candidate = (this.offset + this.coprime * i) % this.candidateCount;
+            final int relX = candidate % (this.radius * 2 + 1);
+            final int relZ = candidate / (this.radius * 2 + 1);
+            final int x = this.spawnSuggestion.getX() + relX - this.radius;
+            final int z = this.spawnSuggestion.getZ() + relZ - this.radius;
+
+            if (this.scheduleCandidate(x, z, i, () -> {
+                final BlockPos pos = getOverworldRespawnPos(this.level, x, z);
+                return pos != null && noCollisionNoLiquid(this.level, pos) ? Optional.of(Vec3.atBottomCenterOf(pos)) : Optional.empty();
+            })) {
+                return;
+            }
+        }
+
+        if (this.finishedFuture.isDone()) {
+            return;
+        }
+
+        scheduleCandidate(this.spawnSuggestion.getX(), this.spawnSuggestion.getZ(), i, () -> Optional.of(PlayerSpawnFinderMixin.fixupSpawnHeight(this.level, this.spawnSuggestion)));
+    }
+
+    private boolean scheduleCandidate(int x, int z, int index, Supplier<Optional<Vec3>> supplier) {
+        int chunkX = SectionPos.blockToSectionCoord(x);
+        int chunkZ = SectionPos.blockToSectionCoord(z);
+        final CompletableFuture<?> future = this.level.getChunkSource().addTicketAndLoadWithRadius(TicketType.SPAWN_SEARCH, new ChunkPos(chunkX, chunkZ), 0);
+        final MinecraftServer server = this.level.getServer();
+
+        if (future.isDone() && server.isSameThread()) {
+            try {
+                future.get();
+                supplier.get().ifPresent(this.finishedFuture::complete);
+            } catch (Throwable error) {
+                this.crash(x, z, index, error);
+            }
+            return false; // stay on loop
+        }
+
+        future.whenCompleteAsync((chunk, error) -> {
+            if (error == null) {
+                try {
+                    Optional<Vec3> result = supplier.get();
+                    if (result.isPresent()) {
+                        this.finishedFuture.complete(result.get());
+                    } else {
+                        this.scheduleNext();
+                    }
+                } catch (Exception exception) {
+                    error = exception;
+                }
+            }
+
+            if (error != null) {
+                this.crash(x, z, index, error);
+            }
+        }, server);
+        return true; // exit loop
```

Comment:
Might want to handle the race, too?
```suggestion
        final AtomicReference<OperationState> state = new AtomicReference<>(OperationState.UNINITIALIZED);
        future.whenCompleteAsync((chunk, error) -> {
            if (error == null) {
                try {
                    Optional<Vec3> result = supplier.get();
                    if (result.isPresent()) {
                        this.finishedFuture.complete(result.get());
                    } else if (state.getAndSet(OperationState.EXECUTED) == OperationState.WAITING) {
                        this.scheduleNext();
                    }
                } catch (Exception exception) {
                    error = exception;
                }
            }

            if (error != null) {
                this.crash(x, z, index, error);
            }
        }, server);
        
        if (state.compareAndSet(OperationState.UNINITIALIZED, OperationState.WAITING)) {
            return true; // exit loop
        }
        
        return false;
```

Reply:
I don't think it's necessary to stay on the loop when the future has already completed. Yes it's a unnecessary recursion but this one doesn't overflow the stack since it's a recursion between different threads.

Reply:
> since it's a recursion between different threads.

Not entirely sure which different thread you are referring to? This is just a classical race condition, the `future#isDone` returns 'false' on the line just above and then it happens to complete before or while executing `future#whenCompleteAsync` and we are still in the same thread.

But yeah, I don't have strong feelings about it too, just wanted to point it out.